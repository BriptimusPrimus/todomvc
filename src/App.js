/* eslint-disable react/jsx-filename-extension */
// Replace ### + jsx with @ + jsx to render with React
/** ###jsx React.createElement */
import React from 'react';
import { dom } from '../todomvc/lib/state-manager/impl1';

const App = () => {
  return (
    <div>
      <p>React here!</p>
    </div>
  );
};
export default App;
