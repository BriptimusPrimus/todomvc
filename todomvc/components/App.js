import { createStore, register } from '../lib/state-ui-lib';
import Main from './Main';
import { app as reducer } from '../reducers';
import controllersFactory from '../controllers';

const App = function App({ initialState, callback }) {
  // Returns a store with a dispatch function to
  // trigger state changes by dispatching actions.
  const store = createStore(initialState, reducer);

  const { onTodoClick, onRemoveClick, onAddButtonClick } = controllersFactory(
    store
  );

  function view(state) {
    return Main({
      todos: state.todos,
      onTodoClick,
      onRemoveClick,
      onAddButtonClick
    });
  }

  // In order to make a component stateful, we must
  // register the rendering function (view) with the store.
  const component = register({
    view,
    store,
    callback
  });

  // All component functions (stateless or stateful)
  // must return the root component node
  return component;
};

export default App;
