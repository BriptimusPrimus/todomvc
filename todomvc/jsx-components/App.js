import { dom, useReducer } from '../lib/state-ui-lib';
import Main from './Main';
import { app as reducer } from '../reducers';
import { toggleTodo, removeTodo, addTodo } from '../actions';

let count = 0;

const App = function App({ initialState, callback }) {
  // In order to make a component stateful, we must
  // register the rendering function (view) with the inner store.
  const [createNode, dispatch] = useReducer({
    initialState,
    reducer,
    callback
  });

  const onTodoClick = id => {
    if (!id) {
      return;
    }
    dispatch(toggleTodo(id));
  };

  const onRemoveClick = id => {
    if (!id) {
      return;
    }
    dispatch(removeTodo(id));
  };

  const onAddButtonClick = text => {
    if (!text) {
      return;
    }
    count += 1;
    dispatch(
      addTodo({
        id: count,
        text
      })
    );
  };

  const view = state => {
    return (
      <Main
        todos={state.todos}
        onTodoClick={onTodoClick}
        onRemoveClick={onRemoveClick}
        onAddButtonClick={onAddButtonClick}
      />
    );
  };

  // All component functions (stateless or stateful)
  // must return the root component node
  return createNode(view);
};

export default App;
