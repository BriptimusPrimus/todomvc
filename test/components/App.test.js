import {
  getByText,
  getByLabelText,
  queryByText,
  getByTestId
} from '@testing-library/dom';

import App from '../../todomvc/components/App';
import { render, cleanup } from '../utils';

describe('App test', () => {
  let mockProps = {};

  beforeEach(() => {
    mockProps = {
      initialState: {
        todos: [
          {
            id: 101,
            text: 'lint with eslint',
            done: true
          },
          {
            id: 102,
            text: 'format with prettier',
            done: true
          },
          {
            id: 103,
            text: 'test with jest',
            done: false
          },
          {
            id: 104,
            text: 'hook with huskey and lint-staged',
            done: false
          }
        ],
        filer: 'all'
      },
      callback: () => {}
    };
  });

  afterEach(cleanup);

  test('Should render', () => {
    const container = render(App(mockProps));
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Add todo', () => {
    const container = render(App(mockProps));
    const newTodo = 'build with webpack';

    const input = getByLabelText(container, 'Task:');
    input.value = newTodo;
    getByText(container, 'Add Todo').click();

    expect(getByText(container, newTodo)).toBeTruthy();
  });

  test('Clik add button without any text', () => {
    const container = render(App(mockProps));

    const todos = container.querySelectorAll('#main ul li');
    getByText(container, 'Add Todo').click();

    // The number of todos is still the same
    expect(container.querySelectorAll('#main ul li').length).toEqual(
      todos.length
    );
  });

  test('Toggle todo', () => {
    let todoItem;
    const container = render(App(mockProps));

    todoItem = getByText(container, 'test with jest');
    expect(todoItem.style.cssText).toEqual('');

    todoItem.click();
    todoItem = getByText(container, 'test with jest');
    expect(todoItem.style.cssText).toEqual('text-decoration: line-through;');
  });

  test('Remove todo', () => {
    const container = render(App(mockProps));

    const todo = mockProps.initialState.todos[0];
    expect(queryByText(container, todo.text)).toBeTruthy();

    const removeLink = getByTestId(container, `remove-${todo.id}`);
    removeLink.click();
    expect(queryByText(container, todo.text)).toBeNull();
  });
});
