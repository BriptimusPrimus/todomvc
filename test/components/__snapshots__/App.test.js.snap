// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test Should render 1`] = `
<div
  id="main"
  style="margin:100px 10px 50px 50px;"
>
  <ul
    style="list-style:none;"
  >
    <li
      id="101"
      style="text-align:right;padding:5px;margin:2px;max-width:200px;cursor:pointer;"
    >
      <span
        style="text-decoration:line-through;"
      >
        lint with eslint
      </span>
      <a
        data-testid="remove-101"
        href="#"
        style="margin-left:20px;"
      >
        X
      </a>
    </li>
    <li
      id="102"
      style="text-align:right;padding:5px;margin:2px;max-width:200px;cursor:pointer;"
    >
      <span
        style="text-decoration:line-through;"
      >
        format with prettier
      </span>
      <a
        data-testid="remove-102"
        href="#"
        style="margin-left:20px;"
      >
        X
      </a>
    </li>
    <li
      id="103"
      style="text-align:right;padding:5px;margin:2px;max-width:200px;cursor:pointer;"
    >
      <span>
        test with jest
      </span>
      <a
        data-testid="remove-103"
        href="#"
        style="margin-left:20px;"
      >
        X
      </a>
    </li>
    <li
      id="104"
      style="text-align:right;padding:5px;margin:2px;max-width:200px;cursor:pointer;"
    >
      <span>
        hook with huskey and lint-staged
      </span>
      <a
        data-testid="remove-104"
        href="#"
        style="margin-left:20px;"
      >
        X
      </a>
    </li>
  </ul>
  <form>
    <label
      for="todo-description"
    >
      Task: 
    </label>
    <input
      id="todo-description"
      placeholder="Todo description"
      type="text"
      value=""
    />
    <button
      id="add-todo"
      style="margin-left:10px;"
    >
      Add Todo
    </button>
  </form>
</div>
`;
